highfidelity/hifi
# Draggable numbers - fix undo, particle angles, being disabled, more robust dragging

Addresses:
https://highfidelity.manuscript.com/f/cases/19978/
https://highfidelity.manuscript.com/f/cases/19997/
https://highfidelity.manuscript.com/f/cases/19998/
https://highfidelity.manuscript.com/f/cases/20085/

Also fixes draggable numbers not being disabled when properties are all disabled, and incorrect label on Spin property.

Pre-Merge Test Plan:
1. Enter Interface in desktop mode into any non-sandbox domain with edit permissions
2. Open Create and create or select any Particle entity
3. Verify if quickly clicking and dragging on vector properties such as Position or Rotation and releasing when over a neighboring value (such as in bug 19978 above) that the neighboring value does not get focused (text highlighted and arrows disappear). This should probably be tested many times as it was hard to repro.
4. Verify the Horizontal Angle Start and Finish properties have ranges of 0 to 180 and the Vertical Angle Start and Finish properties have a range of -180 to 180. Verify that editing all 4 values by dragging them works as expected.
5. Verify if quickly clicking and dragging on numerical particle properties (such as Spin or Size) to perform a very fast change that upon releasing the displayed value stays the same. Also verify if dragging numerical particle properties back and forth very quickly for a few seconds that upon releasing the displayed value stays the same. 
**Mac Note:** It is expected behavior on Mac that some numerical particle properties might lag in this case where the value changes back and forth quickly after releasing (especially if framerate is low), and this behavior is the same as in RC75 on Mac with the old sliders for these values.
6. Perform step 5 again but instead on numerical non-particle properties (such as Dimensions or Local Position).
**Mac Note:** Unlike in step 5, the non-particle properties should not have the delayed update after releasing on Mac.
7. Perform many subsequent drags of multiple different numerical properties, and verify that each undo action will undo each individual value change from drag start to drag finish, and verify that each redo performs as expected.
**Mac Note:** It is expected behavior on Mac that multiple undo actions can be triggered if Cmd + Z keys are held down. This PR, however, should fix the fact that an undo was being triggered on both pressing Cmd + Z and releasing Cmd + Z.
8. Create any new entity and repeat step 7 on the new entity including undoing changes all the way up to deleting the entity.
9. Select a locked entity or lock an entity and verify that attempting to use the arrows on any draggable numbers or clicking to drag them does nothing.
